redaxo/src/core/lib/csrf_manager.php

```diff
@@ -0,0 +1,110 @@
+<?php
+
+/**
+ * @author gharlan
+ *
+ * @package redaxo\core
+ */
+class rex_csrf_manager
+{
+    const PARAM = '_token';
+
+    /**
+     * @param string $tokenId
+     *
+     * @return string
+     */
+    public static function getToken($tokenId)
+    {
+        $tokens = self::getTokens();
+
+        if (isset($tokens[$tokenId])) {
+            return $tokens[$tokenId];
+        }
+
+        $token = self::generateToken();
+        $tokens[$tokenId] = $token;
+        rex_set_session(self::getSessionKey(), $tokens);
+
+        return $token;
+    }
+
+    /**
+     * @param string $tokenId
+     *
+     * @return string
+     */
+    public static function getHiddenField($tokenId)
+    {
+        return sprintf('<input type="hidden" name="%s" value="%s"/>', self::PARAM, self::getToken($tokenId));
+    }
+
+    /**
+     * @param string      $tokenId
+     * @param null|string $token   If `null`, the token is fetched via "_token" request param.
+     *
+     * @return bool
+     */
+    public static function isValid($tokenId, $token = null)
+    {
+        $tokens = self::getTokens();
+
+        if (!isset($tokens[$tokenId])) {
+            return false;
+        }
+
+        if (null === $token) {
+            $token = rex_request(self::PARAM, 'string');
+        }
+
+        if (function_exists('hash_equals')) {
+            return hash_equals($tokens[$tokenId], $token);
+        }
+
+        return $token === $tokens[$tokenId];
+    }
+
+    /**
+     * @param string $tokenId
+     */
+    public static function removeToken($tokenId)
+    {
+        $tokens = self::getTokens();
+
+        unset($tokens[$tokenId]);
+
+        rex_set_session(self::getSessionKey(), $tokens);
+    }
+
+    public static function removeAll()
+    {
+        rex_login::startSession();
+
+        rex_unset_session(self::getSessionKey());
+    }
+
+    private static function getTokens()
+    {
+        rex_login::startSession();
+
+        return rex_session(self::getSessionKey(), 'array');
+    }
+
+    private static function getSessionKey()
+    {
+        return 'csrf_tokens_'.rex::getEnvironment();
+    }
+
+    private static function generateToken()
+    {
+        if (function_exists('random_bytes')) {
```

Comment:
vllt stattdessn besser nen symfony polyfill?

Reply:
Die richtige Polyfill-Lib wäre diese: https://github.com/paragonie/random_compat
(Die wird auch in symfony/polyfill-php70 genutzt.)

Ich kann nicht ganz einschätzen, wie kritisch das an dieser Stelle ist. Mir kam es unverhältnismäßig vor, die nur deswegen reinzunehmen.
Aber wenn du das anders einschätzt, nehme ich die rein.

Reply:
hab grad mal nachgefragt und folgende Antwort bekommen

> hi guys. short question: is a strong hash function important for CSRF tokens? 
or is `uniqid(mt_rand(), true);` enough?

> No, but proper random is important, don't use uniqid.
Just use bin2hex(random_bytes(16));

daher würd ich den polyfill nehmen

Reply:
Ok, dann nehme ich für `hash_equals` (s.u.) symfony/polyfill-php56 rein.
Und würdest du für `random_bytes` dann paragonie/random_compat reinnehmen, da wir das aktuell nur brauchen, oder dann gleich direkt auch das ganze symfony/polyfill-php70 Paket?
https://github.com/symfony/polyfill/tree/master/src/Php70

Reply:
ich würde den ganzen php7 nehmen, da nicht besonders groß